### PR TITLE
Add OutputSequenceSCC

### PR DESCRIPTION
**Pull request by a11ce**

### CODE REVIEW

File: src/core/loop_induction_variables/include/noelle/core/IVStepperUtility.hpp [hunk 1]
```diff
@@ -41,6 +41,12 @@ class IVUtility {
       PHINode *chunkPHI,
       Value *chunkStepSize);
 
+  /*
+   * Finds the isChunkCompleted CmpInst for a previously created chunkPHI
+   */
+  static CmpInst *isChunkCompleted(BasicBlock *preheaderBlock,
```
Comment (scampanoni) [inline src/core/loop_induction_variables/include/noelle/core/IVStepperUtility.hpp:47]: I love it!

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+class OutputSequenceSCC : public LoopCarriedSCC {
+public:
+  OutputSequenceSCC(
+      SCC *s,
+      LoopStructure *loop,
+      const std::set<DGEdge<Value, Value> *> &loopCarriedDependences,
+      const std::set<CallInst *> &outputInstructions);
+
+  std::set<CallInst *> getOutputInstructions(void) const;
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:37]: Can there be instructions within the SCC that are not included in the set returned?

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+  static std::string replacementName(const std::string &originalName,
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:39]: I don't think this method belongs to this abstraction.
It seems to me that this method should be in Gino as the function you'll invoke is part of the Gino's runtime and it is specific to Gino. Another transformation might call a different function.

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                                     bool knownMaxLength);
+
+  static std::optional<int> printConstantFstringMaxLength(CallInst *);
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:42]: Can we find a more descriptive name?

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  static std::optional<std::string> printConstantFstring(CallInst *);
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:43]: Can we find a more descriptive name?

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  static std::optional<int> fstringMaxLength(std::string);
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:44]: Can we find a more descriptive name? 

Does this method belong here or in Gino?

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace arcana::noelle {
+
+class OutputSequenceSCC : public LoopCarriedSCC {
+public:
+  OutputSequenceSCC(
+      SCC *s,
+      LoopStructure *loop,
+      const std::set<DGEdge<Value, Value> *> &loopCarriedDependences,
+      const std::set<CallInst *> &outputInstructions);
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:35]: Why isn't this set automatically computed by the constructor?
Perhaps we can have a static method in OutputSequenceSCC that computes it from an SCC and it is invoked by both SCCDAGAttrs and the OutputSequenceSCC constructor.

**Review by scampanoni — CHANGES_REQUESTED** [review]

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 2 above)
Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:37]: I don't think so. What would replace this, iterateOverInstructions?

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 3 above)
Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:39]: Agreed, along with maybe the fstring-related functions. Where would these go? They're specific to this SCC and don't really operate on the code per se. Would it be a gino tool containing utility functions?

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 6 above)
Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:44]: These should be something like maxPrintedLengthOfFstringUsedByCall, fstringUsedByCall, and maxPrintedLengthOfFstring. But I don't love those for clarity or conveying that they return empty values when not applicable. 

I think these belong here, they probably enable optional space optimizations for any user and don't have any GINO-specific behavior.

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 7 above)
Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:35]: I think I was basing it off of other SCC classifications in SCCDAGAttrs. Why would SCCDAGAttrs and the constructor both need to compute this? SCCDAGAttrs already computes it to classify the SCC.

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 2 above)
Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:37]: Per IRL: Will replace with SCC:getInstructions

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 3 above)
Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:39]: Per IRL: Rename to `parallelReplacementNameForInputOutputFunction` and move to ParallelizationTechnique

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 6 above)
Comment (scampanoni) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:44]: getMaxNumberOfPrintedBytes

Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:44]: Per IRL: Move helpers to protected

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 7 above)
Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:35]: Per IRL: Remove outputInstructions

**Comment by a11ce** [comment]
All current requested changes are made.

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+namespace arcana::noelle {
+
+class OutputSequenceSCC : public LoopCarriedSCC {
+public:
+  OutputSequenceSCC(
+      SCC *s,
+      LoopStructure *loop,
+      const std::set<DGEdge<Value, Value> *> &loopCarriedDependences);
+
+  static std::optional<int> getMaxNumberOfPrintedBytes(CallInst *);
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp:36]: Please avoid using ints with unspecified bitwidths. Please use int32_t and alike.

**Review by scampanoni — CHANGES_REQUESTED** [review]

File: src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp (hunk 8 above)
Comment (a11ce) [reply src/core/loop_scc_attributes/include/noelle/core/OutputSequenceSCC.hpp]: Changed.

**Pull request closed without merging**